### PR TITLE
MGMT-7639: Support using podman in the assisted-service

### PR DESCRIPTION
**Pull request by michaellevy101**
[MGMT-7639](https://issues.redhat.com/browse/MGMT-7639): Support using podman in the deploy-test make target
1. Upgrade our build image form Centos7  to quay.io/centos/centos:stream8 which is relatively new centos because the old image has a very old podman which lack a lot of features
2. Support podman in the make file
3. Run podman-remote from the skipper container by sharing the socket (you have to enable the socket first)
To enable the socket use "systemctl --user enable --now podman.socket" and "sudo loginctl enable-linger $USER"
See: https://github.com/containers/podman/blob/main/docs/tutorials/remote_client.md for details

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @YuviGold 
/cc @ybettan 

## Checklist

- [x] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by maorfr** [comment]
[test]

**Comment by ybettan** [comment]
Thank you for that PR.

What about other targets? `unit-tests` for example? Do you intend to implement it in separate PRs?

File: Dockerfile.assisted-service-build-podman [hunk 1]
```diff
@@ -0,0 +1,19 @@
+FROM quay.io/podman/stable:latest
+
+ENV GO111MODULE=on
+ENV GOFLAGS=""
+ENV CONTAINER_COMMAND="podman"
+ENV VERSION=1.16     GOCACHE=/go/.cache     GOARM=5     GOPATH=/go     GOROOT=/usr/local/go     GOFLAGS='-mod=vendor'     LOGNAME=deadbeef
+ENV PATH=$PATH:$GOROOT/bin:$GOPATH/bin
+
+RUN dnf install -y libguestfs-tools libvirt-client python3 python3-devel unzip gcc git hostname jq krb5-devel libvirt-devel make nmap-ncat openssl rsync tar wget which zip
+RUN set -x && curl --fail --retry 3 -L https://golang.org/dl/go$VERSION.linux-amd64.tar.gz | tar -C /usr/local -xzf - && go get golang.org/x/tools/cmd/cover golang.org/x/tools/cmd/goimports github.com/tools/godep golang.org/x/lint/golint github.com/openshift/release/tools/gotest2junit github.com/openshift/imagebuilder/cmd/imagebuilder && GO111MODULE=on go get gotest.tools/gotestsum@v0.6.0 && mv $GOPATH/bin/* /usr/bin/ && rm -rf $GOPATH/* $GOPATH/.cache && mkdir $GOPATH/bin && ln -s /usr/bin/imagebuilder $GOPATH/bin/imagebuilder && ln -s /usr/bin/goimports $GOPATH/bin/goimports && curl -L https://github.com/golang/dep/releases/download/v0.5.4/dep-linux-amd64 > /usr/bin/dep && chmod +x /usr/bin/dep
```
Comment (ybettan) [inline Dockerfile.assisted-service-build-podman:10]: Why not just:
* `dnf install -y golang`
* setting GOPATH and adding it to PATH
?

Comment (michaellevy101) [reply Dockerfile.assisted-service-build-podman]: irrelevant, new solution, it wasn't ready than

File: Makefile [hunk 2]
```diff
@@ -15,7 +15,7 @@ ifdef E2E_TESTS_MODE
 E2E_TESTS_CONFIG = --img-expr-time=5m --img-expr-interval=5m
 endif
 
-ifeq ($(CONTAINER_COMMAND), podman)
+ifneq (,$(findstring podman,$(CONTAINER_COMMAND)))
```
Comment (ybettan) [inline Makefile:18]: Does that mean that `podman` is chosen if both `docker` and `podman` exist on the machine?

File: hack/utils.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+   # The SKIPPER_UID environment variable is an indication that we are running on a skipper container.
+   [ -n "${SKIPPER_UID+x}" ]
+}
+
+function get_container_runtime_command() {
+    if running_from_skipper; then
+        # skipper creates the SKIPPER_DOCKER_GID env variable for a docker based build
+        CONTAINER_TOOL=$( [[ "$CONTAINER_RUNTIME_COMMAND" == ""docker"" ]] && echo "docker" || echo "podman-remote")
+    else 
+        CONTAINER_TOOL=$( command -v docker &> /dev/null && echo "docker" || echo "podman")
```
Comment (ybettan) [inline hack/utils.sh:14]: In Makefile line 18 seems like `podman` is chosen over docker if both exist, but here seems like it is the opposite?

I don't think it makes a big change but may be worth staying consistent.

File: hack/setup_env.sh [hunk 4]
```diff
@@ -40,11 +40,13 @@ function spectral() {
...
 
 function assisted_service() {
+  mkdir $GOPATH
   latest_kubectl_version=$(curl --retry 5 -L -s https://dl.k8s.io/release/stable.txt)
   curl --retry 5 -L "https://dl.k8s.io/release/${latest_kubectl_version}/bin/linux/amd64/kubectl" -o /tmp/kubectl && \
     install -o root -g root -m 0755 /tmp/kubectl /usr/local/bin/kubectl && \
     rm -f /tmp/kubectl
-  yum install -y docker podman libvirt-clients awscli python3-pip postgresql genisoimage skopeo p7zip && \
+  
+  dnf -y update && dnf install -y golang which jq docker python3 podman-remote libvirt-client awscli python3-pip postgresql genisoimage skopeo p7zip && \
```
Comment (ybettan) [inline hack/setup_env.sh:49]: Isn't `docker` conflicting with the `podman` image here?

File: hack/setup_env.sh [hunk 5]
```diff
@@ -40,11 +40,13 @@ function spectral() {
 }
 
 function assisted_service() {
```
Comment (ybettan) [inline hack/setup_env.sh:42]: Who is using it? I see it only on `init` makefile target but I don't see anyone using that target.

File: hack/utils.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    fi
+
+    # default to docker
+    CONTAINER_TOOL="${CONTAINER_TOOL:-docker}"
```
Comment (ybettan) [inline hack/utils.sh:18]: same

File: hack/setup_env.sh (hunk 4 above)
Comment (YuviGold) [inline hack/setup_env.sh:49]: What golang/python versions are those?

File: hack/setup_env.sh (hunk 5 above)
Comment (YuviGold) [reply hack/setup_env.sh:42]: OpenShift CI

File: hack/setup_env.sh [hunk 7]
```diff
@@ -71,6 +73,7 @@ function assisted_service() {
     github.com/AlekSi/gocov-xml@v0.0.0-20190121064608-3a14fb1c4737
 
   python3 -m pip install --upgrade pip
+  python3 -m venv $VIRTUAL_ENV
```
Comment (YuviGold) [inline hack/setup_env.sh:76]:
```suggestion
  python3 -m venv ${VIRTUAL_ENV}
```

File: hack/setup_env.sh [hunk 8]
```diff
@@ -40,11 +40,13 @@ function spectral() {
...
 function assisted_service() {
+  mkdir $GOPATH
   latest_kubectl_version=$(curl --retry 5 -L -s https://dl.k8s.io/release/stable.txt)
   curl --retry 5 -L "https://dl.k8s.io/release/${latest_kubectl_version}/bin/linux/amd64/kubectl" -o /tmp/kubectl && \
     install -o root -g root -m 0755 /tmp/kubectl /usr/local/bin/kubectl && \
     rm -f /tmp/kubectl
-  yum install -y docker podman libvirt-clients awscli python3-pip postgresql genisoimage skopeo p7zip && \
+  
+  dnf -y update && dnf install -y golang which jq docker python3 podman-remote libvirt-client awscli python3-pip postgresql genisoimage skopeo p7zip && \
     yum clean all
```
Comment (YuviGold) [inline hack/setup_env.sh:50]: dnf clean

File: Dockerfile.assisted-service-build [hunk 9]
```diff
@@ -1,7 +1,9 @@
-FROM registry.ci.openshift.org/openshift/release:golang-1.16
-
+FROM quay.io/podman/stable:latest
```
Comment (YuviGold) [inline Dockerfile.assisted-service-build:1]: Use a fixed version
```suggestion
FROM quay.io/podman/stable:v3.3.0
```

File: skipper.yaml [hunk 10]
```diff
@@ -15,6 +15,9 @@ volumes:
   - $HOME/.docker/:$HOME/.docker
   - $KUBECONFIG:$KUBECONFIG
 
+  # podman - sharing the podman.socket between the host and the skipper container.
+  - /run/user/${UID}/podman/podman.sock:/run/podman/podman.sock
```
Comment (YuviGold) [inline skipper.yaml:19]: Shouldn't it be part of skipper internals? Just like we don't pass the docker socket explicitly.

File: hack/utils.sh [hunk 11]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+set -x
+
+function running_from_skipper() {
+   # The SKIPPER_UID environment variable is an indication that we are running on a skipper container.
+   [ -n "${SKIPPER_UID+x}" ]
```
Comment (YuviGold) [inline hack/utils.sh:6]: We check `SKIPPER_USERNAME` on the Makefile. Could you align them? I don't care which one :) 
https://github.com/openshift/assisted-service/blob/da115d03cbd81b752a3777f4bd5d34459beaf089/Makefile#L127-L130

File: Makefile (hunk 2 above)
Comment (michaellevy101) [reply Makefile:18]: no,  this section is for local registry and tells podman to push even if the registry communicates over http (unsecured).
We have two different podman tools: podman for the host and podman-remote (can use podman --remote but uses podman-remote command for readability) for the skipper container.
This condition checks if the CONTAINER_COMMAND contains the word "podman" so for docker the value will be false

File: hack/utils.sh [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+function get_container_runtime_command() {
+    if running_from_skipper; then
+        # skipper creates the SKIPPER_DOCKER_GID env variable for a docker based build
+        CONTAINER_TOOL=$( [[ "$CONTAINER_RUNTIME_COMMAND" == ""docker"" ]] && echo "docker" || echo "podman-remote")
```
Comment (YuviGold) [inline hack/utils.sh:12]: Did you intend to check the `SKIPPER_DOCKER_GID`? Otherwise, I don't understand this comment.

File: hack/utils.sh (hunk 3 above)
Comment (michaellevy101) [reply hack/utils.sh:14]: please read my comment there

File: hack/setup_env.sh (hunk 4 above)
Comment (michaellevy101) [reply hack/setup_env.sh:49]: no, it's tested

File: hack/utils.sh (hunk 6 above)
Comment (YuviGold) [inline hack/utils.sh:18]: This line can be removed as `CONTAINER_TOOL` would always be assigned above.

File: hack/setup_env.sh (hunk 5 above)
Comment (michaellevy101) [reply hack/setup_env.sh:42]: The "Dockerfile.assisted-service-build"

File: hack/setup_env.sh (hunk 4 above)
Comment (michaellevy101) [reply hack/setup_env.sh:49]: good point, currently Python 3.9.6 and go 1.16.6

File: skipper.yaml (hunk 10 above)
Comment (michaellevy101) [reply skipper.yaml:19]: not sure, it's unsecured so need to think about it carefully
Let's start with it this way
please read this thread: https://coreos.slack.com/archives/CBBJY9GSX/p1629391006061200

File: hack/utils.sh (hunk 12 above)
Comment (michaellevy101) [reply hack/utils.sh:12]: sorry:) old junk

File: hack/utils.sh (hunk 6 above)
Comment (michaellevy101) [reply hack/utils.sh]: this line removed because it's redundant

File: hack/setup_env.sh (hunk 4 above)
Comment (YuviGold) [reply hack/setup_env.sh:49]: Can it be fixed somehow?

File: skipper.yaml (hunk 10 above)
Comment (YuviGold) [reply skipper.yaml:19]: It's the same amount of insecure with docker. I think it should be aligned on the skipper level as the skipper wasn't created as a security tool but a developer-friendly one. That runs as privileged with a host network, user info, etc.
@eranco74 WDYT?

File: hack/utils.sh [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+set -x
+
+function running_from_skipper() {
+   # The SKIPPER_UID environment variable is an indication that we are running on a skipper container.
+   [ -n "${SKIPPER_UID+x}" ]
+}
+
+function get_container_runtime_command() {
+    if running_from_skipper; then
+        CONTAINER_TOOL=$( [[ "$CONTAINER_RUNTIME_COMMAND" == ""docker"" ]] && echo "docker" || echo "podman-remote")
```
Comment (YuviGold) [inline hack/utils.sh:11]:
```suggestion
        CONTAINER_TOOL=$( [[ "${CONTAINER_RUNTIME_COMMAND}" == ""docker"" ]] && echo "docker" || echo "podman-remote")
```

File: hack/setup_env.sh (hunk 4 above)
Comment (michaellevy101) [reply hack/setup_env.sh:49]: need to investigate

File: skipper.yaml (hunk 10 above)
Comment (michaellevy101) [reply skipper.yaml:19]: @YuviGold I agree that for administrator users it's reasonable but unlike docker you have to enable the socket (disabled by default) and need to think about  solution for unprivileged users (docker runs as root but podman isn't).
For now let continue without changing skipper.

Comment (eranco74) [reply skipper.yaml:19]: I'm fine with both options.
Let's start with this

File: Dockerfile.assisted-service-build [hunk 14]
```diff
@@ -1,12 +1,14 @@
...
-
+FROM quay.io/centos/centos:stream8
 ENV GO111MODULE=on
 ENV GOFLAGS=""
+ENV VIRTUAL_ENV=/opt/venv
+ENV GOPATH=/go
+ENV PATH="$VIRTUAL_ENV/bin:$GOPATH/bin:$PATH"
 
 COPY ./hack/setup_env.sh ./dev-requirements.txt ./
+RUN dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm && dnf -y update && dnf -y install make git jq golang python3 && dnf clean all
```
Comment (YuviGold) [inline Dockerfile.assisted-service-build:9]: Is that necessary to install the tools in here? I see it is installed as part of `setup_env`

Comment (YuviGold) [inline Dockerfile.assisted-service-build:9]: Add comment why we need this `epel`

File: hack/utils.sh [hunk 15]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+function running_from_skipper() {
+   # The SKIPPER_UID environment variable is an indication that we are running on a skipper container.
+   [ -n "${SKIPPER_UID+x}" ]
+}
+
+function get_container_runtime_command() {
+    if running_from_skipper; then
+        CONTAINER_TOOL=$( [[ "${CONTAINER_RUNTIME_COMMAND}" == ""docker"" ]] && echo "docker" || echo "podman-remote")
```
Comment (YuviGold) [inline hack/utils.sh:10]:
```suggestion
        CONTAINER_TOOL=$( [[ "${CONTAINER_RUNTIME_COMMAND}" == "docker" ]] && echo "docker" || echo "podman-remote")    
```

File: skipper.yaml [hunk 16]
```diff
@@ -15,6 +15,10 @@ volumes:
   - $HOME/.docker/:$HOME/.docker
   - $KUBECONFIG:$KUBECONFIG
 
+  # podman - sharing the podman.socket between the host and the skipper container
+  # set USER_ID to $UID manually. TODO: add support to skipper
```
Comment (YuviGold) [inline skipper.yaml:19]: fix comment

File: hack/generate.sh [hunk 17]
```diff
@@ -6,21 +6,22 @@ set -o errexit
 
 __dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 __root="$(cd "$(dirname "${__dir}")" && pwd)"
+source ${__dir}/utils.sh sleep 0
```
Comment (YuviGold) [inline hack/generate.sh:9]: Add comment

**Pull request merged**